core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperator.java

```diff
@@ -0,0 +1,218 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+/**
+ * Group the all the input {@link BindingTuple} by {@link RareTopNOperator#groupByExprList},
+ * Calculate the rare result by using the {@link RareTopNOperator#fieldExprList}.
+ */
+@ToString
+@EqualsAndHashCode
+public class RareTopNOperator extends PhysicalPlan {
+
+  @Getter
+  private final PhysicalPlan input;
+
+  /**
+   * If rareTopFlag is true, result for top will be returned.
+   * If rareTopFlag is false, result for rare will be returned.
+   */
+  @Getter
+  private final Boolean rareTopFlag;
+  @Getter
+  private final Integer noOfResults;
+  @Getter
+  private final List<Expression> fieldExprList;
+  @Getter
+  private final List<Expression> groupByExprList;
+
+  @EqualsAndHashCode.Exclude
+  private final Group group;
+  @EqualsAndHashCode.Exclude
+  private Iterator<ExprValue> iterator;
+
+  private static final Integer DEFAULT_NO_OF_RESULTS = 10;
+
+
+  public RareTopNOperator(PhysicalPlan input, Boolean rareTopFlag, List<Expression> fieldExprList,
+      List<Expression> groupByExprList) {
+    this(input, rareTopFlag, DEFAULT_NO_OF_RESULTS, fieldExprList, groupByExprList);
+  }
+
+  /**
+   * RareTopNOperator Constructor.
+   *
+   * @param input           Input {@link PhysicalPlan}
+   * @param rareTopFlag     Flag for Rare/TopN command.
+   * @param noOfResults     Number of results
+   * @param fieldExprList   List of {@link Expression}
+   * @param groupByExprList List of group by {@link Expression}
+   */
+  public RareTopNOperator(PhysicalPlan input, Boolean rareTopFlag, int noOfResults,
+      List<Expression> fieldExprList,
+      List<Expression> groupByExprList) {
+    this.input = input;
+    this.rareTopFlag = rareTopFlag;
+    this.noOfResults = noOfResults;
+    this.fieldExprList = fieldExprList;
+    this.groupByExprList = groupByExprList;
+    this.group = new Group();
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitRareTopN(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return iterator.hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    return iterator.next();
+  }
+
+  @Override
+  public void open() {
+    super.open();
+    while (input.hasNext()) {
+      group.push(input.next());
+    }
```

Comment:
Should we do something to indicate that input is no longer usable or throw an error if open() is called more than once?

Reply:
I am not sure if it is possible to call open() more than once for a single request. AggregationOperator & SortOperator also override open() and it doesn't throw an error so I assumed it's not required.

Reply:
I would suggest using Precondition.checkNull(iterator) here to fail fast if open() accidentally gets called more than once.

Reply:
open() gets called only once